modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java

```diff
@@ -0,0 +1,358 @@
+            Iterator<CollisionJobContext> iter = collisioinSpi.jobs.values().iterator();
+
+            iter.next().cancel();
+
+            assertEquals(1, rejected.value());
+
+            Thread.sleep(100); //Sleeping to make sure totalWaitingTime will become more the zero.
+
+            iter.next().activate();
+            iter.next().activate();
+
+            boolean res = waitForCondition(() -> active.value() > 0, TIMEOUT);
+
+            assertTrue(res);
+
+            Thread.sleep(100); //Sleeping to make sure totalExecutionTime will become more the zero.
+
+            latch.countDown();
+
+            res = waitForCondition(() -> fut1.isDone() && fut2.isDone() && fut3.isDone(), TIMEOUT);
+
+            assertTrue(res);
+
+            res = waitForCondition(() -> finished.value() == 3, TIMEOUT);
+
+            assertTrue(res);
+
+            assertTrue("Execution time should be greater then zero.", totalExecutionTime.value() > 0);
+            assertTrue("Waiting time should be greater then zero.", totalWaitingTime.value() > 0);
+        }
+    }
+
+    /** Test correct calculation of finished, started, active, canceled metrics of the {@link GridJobProcessor}. */
+    @Test
+    public void testGridJobMetrics() throws Exception {
+        latch = new CountDownLatch(1);
+
+        try(IgniteEx g = startGrid(0)) {
+            ReadOnlyMetricRegistry mreg = g.context().metric().registry().withPrefix(JOBS);
+
+            LongMetric started = (LongMetric)mreg.findMetric(STARTED);
+            LongMetric active = (LongMetric)mreg.findMetric(ACTIVE);
+            LongMetric waiting = (LongMetric)mreg.findMetric(WAITING);
+            LongMetric canceled = (LongMetric)mreg.findMetric(CANCELED);
+            LongMetric rejected = (LongMetric)mreg.findMetric(REJECTED);
+            LongMetric finished = (LongMetric)mreg.findMetric(FINISHED);
+            LongMetric totalExecutionTime = (LongMetric)mreg.findMetric(EXECUTION_TIME);
+            LongMetric totalWaitingTime = (LongMetric)mreg.findMetric(WAITING_TIME);
+
+            assertNotNull(started);
+            assertNotNull(active);
+            assertNotNull(waiting);
+            assertNotNull(canceled);
+            assertNotNull(rejected);
+            assertNotNull(finished);
+            assertNotNull(totalExecutionTime);
+            assertNotNull(totalWaitingTime);
+
+            assertEquals(0, started.value());
+            assertEquals(0, active.value());
+            assertEquals(0, waiting.value());
+            assertEquals(0, canceled.value());
+            assertEquals(0, rejected.value());
+            assertEquals(0, finished.value());
+            assertEquals(0, totalExecutionTime.value());
+            assertEquals(0, totalWaitingTime.value());
+
+            SimplestTask task = new SimplestTask();
+
+            g.compute().execute(task, 1);
+
+            //Waiting task to finish.
+            boolean res = waitForCondition(() -> active.value() == 0, TIMEOUT);
```

Comment:
Is there a race possible? `active == 0` can be in two cases: when the job is not activated yet and after the job is finished. Here you wait for `active == 0` and then checks for `finished == 1`. Does job always activated after `execute()` in this test?

Reply:
In this test `CollisionSpi` is null.
So `GridJobProcessor#jobAlwaysActivate` will be true.
So yes, Ignite will always activate job on `g.compute().execute(task, 1);` invocation.
For more details, please, see `GridJobProcessor#processJobExecuteRequest` line 1268 in this PR and `GridJobProcessor#onBeforeActivateJob` method.